### PR TITLE
Use new style property decorators.

### PR DESCRIPTION
**Pull request by andrewandrepowell**
Hello,

This change is to resolve issue #719.

### CODE REVIEW

File: cocotb/binary.py [hunk 1]
```diff
@@ -376,16 +406,25 @@ def set_binstr(self, string):
...
         """The number of bits of the binary value."""
         return self._n_bits
-
-    n_bits = property(_get_n_bits, None, None,
-                      "Access to the number of bits of the binary value.")
+        
+    @property
+    def n_bits(self):
+        "Access to the number of bits of the binary value."
+        return self._get_n_bits()
```
Comment (eric-wieser) [inline cocotb/binary.py]: Can you just inline this method, since it's private?

**Comment by eric-wieser** [comment]
This is changing <=2.5 style to >=2.6 style, not 2.7 to 3.x!

File: cocotb/binary.py (hunk 1 above)
Comment (andrewandrepowell) [reply cocotb/binary.py]: Done

**Comment by andrewandrepowell** [comment]
Made the requested changes.

File: cocotb/binary.py [hunk 2]
```diff
@@ -303,12 +303,35 @@ def is_resolvable(self):
...
 
-    value = property(get_value, set_value, None,
-                     "Integer access to the value. **deprecated**")
-    integer = property(get_value, set_value, None,
-                       "The integer representation of the underlying vector.")
-    signed_integer = property(get_value_signed, set_value, None,
-                              "The signed integer representation of the underlying vector.")
+    @property
+    def value(self):
+        "Integer access to the value. **deprecated**"
```
Comment (ktbarrett) [inline cocotb/binary.py]: I think we should take this opportunity to deprecate this interface by adding a warning and removing the message from the docstring.

File: cocotb/binary.py [hunk 3]
```diff
@@ -303,12 +303,35 @@ def is_resolvable(self):
...
+        self.set_value(val)
+
+    @property
+    def integer(self):
+        "The integer representation of the underlying vector."
+        return self.get_value()
+
+    @integer.setter
+    def integer(self, val):
+        "The integer representation of the underlying vector."
```
Comment (ktbarrett) [inline cocotb/binary.py]: You can leave off the docstrings on all of the setters, only the getter's docstring is used.

File: cocotb/binary.py [hunk 4]
```diff
@@ -303,12 +303,35 @@ def is_resolvable(self):
...
+        self.set_value(val)
+
+    @property
+    def signed_integer(self):
+        "The signed integer representation of the underlying vector."
+        return self.get_value_signed()
```
Comment (eric-wieser) [inline cocotb/binary.py]: I think we should flip the definitions here, from
```
    @property
    def signed_integer(self):
        "The signed integer representation of the underlying vector."
        return self.get_value_signed()
    def get_value_signed(self):
         ...
```
to
```
    @property
    def signed_integer(self):
        "The signed integer representation of the underlying vector."
        ...

    get_value_signed = signed_integer.fget
```
or similar

Comment (andrewandrepowell) [reply cocotb/binary.py]: Completed this specific request. The other cases are the `set_value` and `get_value` methods. Should the same change be applied to those as well, and I'm assuming their definitions will be used for the integer property.

Comment (eric-wieser) [reply cocotb/binary.py]: Yes please!

**Comment by ktbarrett** [comment]
The warning is throwing, but only on Python 3.8. I figured somewhere the filter was changed to `"error"`, but I don't see that. Sounds like you are for some digging.

**Comment by eric-wieser** [comment]
Yep, only 3.8 let's you configure warnings as errors via an environment var, so 3.8 is the only CI we have that turned on for.

One of our tests or examples is using this deprecated property.

Let's not worry about the deprecation in this PR - mixing deprecations and cleanups is usually a bad idea.

File: cocotb/binary.py [hunk 5]
```diff
@@ -295,22 +319,34 @@ def get_value_signed(self):
...
-    integer = property(get_value, set_value, None,
-                       "The integer representation of the underlying vector.")
-    signed_integer = property(get_value_signed, set_value, None,
-                              "The signed integer representation of the underlying vector.")
+    def get_hex_buff(self):
+        bstr = self.buff
+        hstr = '%0*X' % ((len(bstr) + 3) // 4, int(bstr, 2))
+        return hstr
+
+    def _adjust(self):
```
Comment (eric-wieser) [inline cocotb/binary.py]: Can you try to keep the code in the same order as before to keep the diff small? This function for instance did not change, but you've moved so much code from above it to below it that git thinks it did.

File: cocotb/binary.py [hunk 6]
```diff
@@ -295,22 +319,34 @@ def get_value_signed(self):
         else:
             return -1 * (1 + (int(~ival) & (signbit - 1)))
 
-    def set_value(self, integer):
-        self._str = self._convert_to[self.binaryRepresentation](integer)
+    get_value_signed = signed_integer.fget
 
-    @property
-    def is_resolvable(self):
```
Comment (eric-wieser) [inline cocotb/binary.py]: Similarly, it would be good to put this function back where it was.

File: cocotb/binary.py [hunk 7]
```diff
@@ -332,14 +352,10 @@ def get_buff(self):
                 buff = chr(val) + buff
         return buff
 
-    def get_hex_buff(self):
-        bstr = self.get_buff()
-        hstr = '%0*X' % ((len(bstr) + 3) // 4, int(bstr, 2))
-        return hstr
-
```
Comment (eric-wieser) [inline cocotb/binary.py]: Can you move this function back too?

File: cocotb/binary.py (hunk 2 above)
Comment (eric-wieser) [reply cocotb/binary.py]: Tracked by #1503, marking as resolved here.

File: cocotb/binary.py (hunk 7 above)
Comment (eric-wieser) [reply cocotb/binary.py]: (the change from `self.get_buff()` to `self.buff` is still a good one)

File: cocotb/binary.py [hunk 8]
```diff
@@ -361,35 +377,34 @@ def _adjust(self):
...
-                    "Access to the value as a buffer.")
+    get_buff = buff.fget
+    set_buff = buff.fset
 
-    def get_binstr(self):
-        """Attribute :attr:`binstr` is the binary representation stored as
-        a string of ``1`` and ``0``."""
+    @property
+    def binstr(self):
+        "Access to the binary string."
```
Comment (eric-wieser) [inline cocotb/binary.py]:
```suggestion
        """ The binary representation stored as a string of ``0``, ``1``, and possibly ``x``, ``z``, and other states. """
```

File: cocotb/binary.py [hunk 9]
```diff
@@ -281,12 +281,30 @@ def _adjust_twos_comp(self, x):
...
+    @integer.setter
+    def integer(self, val):
+        self._str = self._convert_to[self.binaryRepresentation](val)
+
+    get_value = integer.fget
+    set_value = integer.fset
+
+    @property
+    def signed_integer(self):
+        "The signed integer representation of the underlying vector."
```
Comment (eric-wieser) [inline cocotb/binary.py]:
```suggestion
        """The signed integer representation of the underlying vector."""
```
While what you have works, it's not conventional to write docstrings with a single set of quotes.

File: cocotb/binary.py [hunk 10]
```diff
@@ -281,12 +281,30 @@ def _adjust_twos_comp(self, x):
...
         return self._convert_from[self.binaryRepresentation](self._str)
 
-    def get_value_signed(self):
-        """Return the signed integer representation of the underlying vector."""
+    @integer.setter
+    def integer(self, val):
+        self._str = self._convert_to[self.binaryRepresentation](val)
+
+    get_value = integer.fget
+    set_value = integer.fset
```
Comment (eric-wieser) [inline cocotb/binary.py]: Would like it to be a little clearer that this is tied to the `value` property, and will be deprecated alongside it if we ever do that.
```suggestion

    get_value = value.fget
    set_value = value.fset
```

File: cocotb/binary.py (hunk 7 above)
Comment (andrewandrepowell) [reply cocotb/binary.py]: Oops. Missed that one!

File: cocotb/binary.py [hunk 11]
```diff
@@ -295,22 +313,19 @@ def get_value_signed(self):
...
 
-    value = property(get_value, set_value, None,
-                     "Integer access to the value. **deprecated**")
-    integer = property(get_value, set_value, None,
-                       "The integer representation of the underlying vector.")
-    signed_integer = property(get_value_signed, set_value, None,
-                              "The signed integer representation of the underlying vector.")
+    @signed_integer.setter
+    def signed_integer(self, val):
+        self.integer = val
```
Comment (eric-wieser) [inline cocotb/binary.py]: Might make sense to put this on line 315 so that the getter and setter are adjacent.

File: cocotb/binary.py [hunk 12]
```diff
@@ -281,12 +281,30 @@ def _adjust_twos_comp(self, x):
...
-        """Return the integer representation of the underlying vector."""
+    @property
+    def value(self):
+        """Integer access to the value. **deprecated**"""
+        return self.integer
+
+    @value.setter
+    def value(self, val):
+        self.integer = val
+
```
Comment (eric-wieser) [inline cocotb/binary.py]: Looking at the docs, this makes this appear before `integer`: https://cocotb--1497.org.readthedocs.build/en/1497/library_reference.html?highlight=binaryvalue#cocotb.binary.BinaryValue.value

Maybe move these lines down to line 301?

File: cocotb/binary.py [hunk 13]
```diff
@@ -281,12 +281,30 @@ def _adjust_twos_comp(self, x):
...
         return self._convert_from[self.binaryRepresentation](self._str)
 
-    def get_value_signed(self):
-        """Return the signed integer representation of the underlying vector."""
+    @integer.setter
+    def integer(self, val):
+        self._str = self._convert_to[self.binaryRepresentation](val)
+
+    get_value = value.fget
+    set_value = value.fset
```
Comment (ktbarrett) [inline cocotb/binary.py]: The nittiest of nitpicks, this should probably be moved up to directly below the `value` property definitions.

Comment (eric-wieser) [reply cocotb/binary.py]: Better to do the reverse and move the property down: https://github.com/cocotb/cocotb/pull/1497#discussion_r395069321

**Comment by eric-wieser** [comment]
If you're feeling generous, can you add `:exclude-members: get_value, get_buff, get_binstr, get_value_signed` to `BinaryValue` [here](https://github.com/cocotb/cocotb/blob/master/documentation/source/library_reference.rst#interacting-with-the-simulator)?

**Comment by eric-wieser** [comment]
Went ahead and made those tweaks myself, will merge once CI passes. Thanks again @andrewandrepowell!

**Comment by eric-wieser** [comment]
Docs look good now: https://cocotb--1497.org.readthedocs.build/en/1497/library_reference.html#cocotb.binary.BinaryValue

**Pull request merged**